mantidproject/mantid
# Convert MdiSubWindow to use dynamic factory

This converts classes derived from MdiSubWindow which are serialised to Mantid project files to be created by a new dynamic factory, drastically simplifying how each object is loaded at a high level. 

Each window type can now be created using a default constructor from the factory. The `loadFromProject` function will then initialise and load setting for each window. The loading logic for each type of window is now contained within the class itself, i.e. MdiSubWindows now completely know how to load themselves from file. 

Unfortunately there is still a large dependance on ApplicationWindow, but reducing that would be getting beyond the scope of this ticket.

This PR should not add any new functionality or fix any bugs, but hopefully should not break anything either.

**To test:**

A lot of code for project loading/saving has been moved around, but on the surface it should appears as if nothing has changed.
- Check all the tests pass. (especially MantidPlotProjectSerialise!)
- Check that all types of windows can be saved and loaded back. For a list of what should and shouldn't work see issue #15602. Types of windows should include:
  - MantidMatrix
  - Matrix
  - Note
  - Table
  - TableStatistics
  - 1D plots
  - 2D plots
  - 3D plots
  - InstrumentWindow
  - ScriptWindow
- Code review.

Fixes #16768.

_Does not need to be in the release notes._

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
##### Functional Tests
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.
